File: src/randomizer/cost-supply-ban.ts
import {Cards} from "../utils/cards";
import type {CostType} from "../dominion/cost-type";
import type {SupplyBan} from "./supply-ban";
import type {SupplyCard} from "../dominion/supply-card";

export class CostSupplyBan implements SupplyBan {
  constructor(readonly costTypes: CostType[]) {
  }

  getBannedCards(cards: SupplyCard[]): SupplyCard[] {
    return cards.filter(Cards.filterByAllowedCosts(this.costTypes));
  }
}

File: flutter/dominion_randomizer/lib/services/randomizer/cost_supply_ban.dart
import 'package:dominion_randomizer/models/card.dart';
import 'package:dominion_randomizer/models/cost.dart';
import 'supply_ban.dart';

class CostSupplyBan implements SupplyBan {
  final Set<CostType> costTypes;

  CostSupplyBan(this.costTypes);

  @override
  Iterable<Card> getBannedCards(Iterable<Card> cards) {
    return cards.where((card) => costTypes.contains(card.cost.getType()));
  }
}
